### PR TITLE
[FEATURE] WIP: Make rigid body simulation fully differentiable

### LINKED ISSUE #2069: [Bug]: No gradients wrt control force
### Bug Description

I am trying to get gradients wrt control inputs, but the gradient graph seems to be broken. I can get grads wrt initial pose (as in the unit test) but not for control.

### Steps to Reproduce

```
import time
import numpy as np
import torch
import genesis as gs
from etils import epath
from torchviz import make_dot

gs.init(logging_level="warning", backend=gs.cuda)

dt = 1e-2
horizon = 10
substeps = 1
num_iter = 400
lr = 5e-2

scene = gs.Scene(
    sim_options=gs.options.SimOptions(
        dt=dt,
        substeps=substeps,
        requires_grad=True,
        gravity=(0.0, 0.0, -9.81),
    ),
    rigid_options=gs.options.RigidOptions(
        enable_collision=False,
        enable_self_collision=False,
        enable_joint_limit=False,
        disable_constraint=False,
        use_contact_island=False,
        use_hibernation=False,
    ),
    viewer_options=gs.options.ViewerOptions(
        camera_pos=(0.0, -4.0, 1.5),
        camera_lookat=(0.0, 0.0, 1.0),
        camera_fov=30.0,
        res=(960, 640),
        max_FPS=240,
    ),
    show_viewer=True,
)

mjcf_path = str(epath.resource_path('genesis_playground') / "envs/xmls/cartpole.xml")

cartpole = scene.add_entity(
    gs.morphs.MJCF(file=mjcf_path),
)

# Joints: slider (cart x), hinge_1 (pole angle)
slider_joint = cartpole.get_joint("slider")
hinge_joint = cartpole.get_joint("hinge_1")

# Local q indices (for generalized coordinates)
slider_q_idx = slider_joint.q_idx_local
hinge_q_idx = hinge_joint.q_idx_local

scene.build()
cartpole.set_dofs_armature([0.0, 0.0], [slider_q_idx, hinge_q_idx])
cartpole.set_dofs_force_range(
    lower=np.array([-10.0, -10.0]),
    upper=np.array([10.0, 10.0]),
    dofs_idx_local=[slider_q_idx, hinge_q_idx]
)


# Learnable initial slider position and hinge angle
q_init = gs.tensor([1.0, 0.5], requires_grad=True)
q = q_init.detach()
q.requires_grad_()
control = gs.tensor([3.0], requires_grad=True)
optimizer = torch.optim.Adam([control, q], lr=lr)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
    optimizer, T_max=num_iter, eta_min=1e-3
)

# Optimization loop
for it in range(num_iter):
    # scene.reset()
    cartpole.set_qpos(q)
    cartpole.set_dofs_velocity(control, slider_q_idx)
    
    # Rollout
    for i in range(horizon):
        scene.step()

    # Final state
    q_final = cartpole.get_qpos()
    x_T = q_final[slider_q_idx]
    theta_T = q_final[hinge_q_idx]

    # Loss: ONLY make pole upright at final time (theta_T -> 0)
    loss = theta_T**2 + x_T**2
    optimizer.zero_grad()


    make_dot(loss, 
             params={
                 "q": q, 
                 "control": control, 
                 "q_final": q_final, 
                 "x_T": x_T, 
                 "theta_T": theta_T,
                 "loss": loss,
                 },
             show_attrs=True,
             show_saved=True).render("autograd_graph", format="png")


    # retain_graph=True so Genesis' internal graph is not freed
    loss.backward(retain_graph=True)
    optimizer.step()
    scheduler.step()


    print(
        f"[iter {it:03d}] "
        f"grad={q.grad}, "
        f"ctrl grad={control.grad}, "
    )
```

### Expected Behavior

Gradients not none

### Screenshots/Videos

Image

### Relevant log output

```shell
[iter 000] grad=tensor([2.0284, 1.0394], device='cuda:0'), ctrl grad=None, 
[iter 001] grad=tensor([1.9287, 0.9310], device='cuda:0'), ctrl grad=None, 
[iter 002] grad=tensor([1.8292, 0.8229], device='cuda:0'), ctrl grad=None, 
[iter 003] grad=tensor([1.7300, 0.7156], device='cuda:0'), ctrl grad=None, 
[iter 004] grad=tensor([1.6313, 0.6097], device='cuda:0'), ctrl grad=None,
```

### Environment

- OS: [e.g. Ubuntu 24.04, Windows 11 24H2]
- GPU/CPU [e.g. A100, RTX 4090, M3pr, Intel I9-9900k, Ryzen 5900x] (N/A if no GPU/CPU)
- GPU-driver version (N/A if no GPU)
- CUDA / CUDA-toolkit version (N/A if non-Nvidia)

### Release version or Commit ID

a2787f5fa1aeac250c4f36702b5de949993ec057

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by SonSang**
## Description

This PR aims at stitching differentiable components in the rigid body simulation to make it fully differentiable end-to-end. Also, it will provide a minimal demo to demonstrate the differentiability of the rigid body simulation.

Specifically, this PR would contain:
* Fully differentiable rigid body simulation
* Minimal demo for differentiable rigid body simulation (e.g. solving optimization problem)
* More informative error handlings for differentiable simulation for users

## Related Issue

Resolves Genesis-Embodied-AI/Genesis#2069

## How Has This Been / Can This Be Tested?

TBD

## Screenshots (if appropriate):

TBD

## Checklist:

- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [ ] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [ ] I tested my changes and added instructions on how to test it for reviewers.

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: genesis/engine/entities/rigid_entity/rigid_entity.py [hunk 1]
```diff
@@ -2330,7 +2330,7 @@ def set_dofs_velocity(self, velocity=None, dofs_idx_local=None, envs_idx=None, *
 
     @gs.assert_built
     def set_dofs_velocity_grad(self, dofs_idx_local, envs_idx, velocity_grad):
-        dofs_idx = self._get_idx(dofs_idx_local, self.n_dofs, self._dof_start, unsafe=True)
+        dofs_idx = self._get_global_idx(dofs_idx_local, self.n_dofs, self._dof_start, unsafe=True)
```
Comment (duburcqa) [inline genesis/engine/entities/rigid_entity/rigid_entity.py:2351]: We need a unit test for this.

File: genesis/engine/solvers/rigid/rigid_solver_decomp.py [hunk 2]
```diff
@@ -299,6 +302,9 @@ def build(self):
             if getattr(self._options, "noslip_iterations", 0) > 0:
                 gs.raise_exception("Noslip is not supported yet when requires_grad is True.")
 
+            if getattr(self._options, "sparse_solve", False):
+                gs.raise_exception("Sparse solve is not supported yet when requires_grad is True.")
+
```
Comment (duburcqa) [inline genesis/engine/solvers/rigid/rigid_solver_decomp.py:307]: It would be nice to have unit tests to catch all this forbidden configurations.

**Comment by duburcqa** [comment]
Superseded by https://github.com/Genesis-Embodied-AI/genesis-world/pull/2842
